### PR TITLE
Add support for PMW3389 optical sensor

### PR DESCRIPTION
**Pull request by Alabastard-64**
## Description
Added PMW3389 driver as some registries have changed from the PMW3360 including the CPI settings (which have changed to 16bit from 8bit) requiring changes to the set and get CPI functions.
Also I removed the comment from the PMW3360 driver suggesting it should work for the PMW3389 and later sensors as that is not quite correct.
All other changes were to support this addition.
Has been tested on a keyboard with a PMW3389 sensor and the full CPI range seems to be working.

Documentation would need to be updated with the PMW3389 sensor but all factors/settings are identical other than the CPI range is 0 - 16000 in steps of 50.

There is one additional feature of the PMW3360 and PMW3389 chips that may be useful to add in the future, which is the ability to have different CPI settings for the X and Y axes, this uses the Config5 register (again this is 8bit for PMW3360 and 16bit for PMW3389) and requires a bit set in the Config2 register so this would need to be set during initialization.

## Types of Changes

- [X] Core
- [ ] Bugfix
- [X] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* Improper support for PMW3389 optical sensor

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [X] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [X] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by Alabastard-64** [comment]
Whew the commits were a bit messy here.  Sorry first time using git.

File: common_features.mk [hunk 1]
```diff
@@ -356,7 +359,7 @@ endif
 ifeq ($(strip $(PRINTING_ENABLE)), yes)
     OPT_DEFS += -DPRINTING_ENABLE
     SRC += $(QUANTUM_DIR)/process_keycode/process_printer.c
-    QUANTUM_LIB_SRC += uart.c
+    SRC += $(TMK_DIR)/protocol/serial_uart.c
```
Comment (drashna) [inline common_features.mk:362]: Please revert this change. 
```suggestion
    QUANTUM_LIB_SRC += uart.c
```

File: common_features.mk [hunk 2]
```diff
@@ -604,7 +607,6 @@ ifeq ($(strip $(UNICODE_ENABLE)), yes)
 endif
 
 ifeq ($(strip $(UNICODE_COMMON)), yes)
-    OPT_DEFS += -DUNICODE_COMMON_ENABLE
```
Comment (drashna) [inline common_features.mk:607]:
```suggestion
    OPT_DEFS += -DUNICODE_COMMON_ENABLE
```

File: quantum/pointing_device_drivers.c [hunk 3]
```diff
@@ -210,7 +210,7 @@ const pointing_device_driver_t pointing_device_driver = {
 // clang-format on
 #elif defined(POINTING_DEVICE_DRIVER_pmw3360)
 
-static void init(void) { pmw3360_init(); }
+static void init(void) { pmw3360_init(); } //What is this line?
```
Comment (drashna) [inline quantum/pointing_device_drivers.c:213]:
```suggestion
static void init(void) { pmw3360_init(); }
```

File: quantum/pointing_device_drivers.c [hunk 4]
```diff
@@ -244,6 +244,45 @@ const pointing_device_driver_t pointing_device_driver = {
     .set_cpi    = pmw3360_set_cpi,
     .get_cpi    = pmw3360_get_cpi
 };
+
+// clang-format on
+#elif defined(POINTING_DEVICE_DRIVER_pmw3389)
+
+static void init(void) { pmw3389_init(); }
```
Comment (drashna) [inline quantum/pointing_device_drivers.c:251]: Thinking of it it, ...

```suggestion
```

File: quantum/pointing_device_drivers.c [hunk 5]
```diff
@@ -244,6 +244,45 @@ const pointing_device_driver_t pointing_device_driver = {
...
+        mouse_report.x = constrain_hid(data.dx);
+        mouse_report.y = constrain_hid(data.dy);
+    }
+
+    return mouse_report;
+}
+
+// clang-format off
+const pointing_device_driver_t pointing_device_driver = {
+    .init       = init,
```
Comment (drashna) [inline quantum/pointing_device_drivers.c:280]:
```suggestion
    .init       = pmw3389_init,
```

File: drivers/sensors/pmw3389.c [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    spi_stop_adv();
+
+    wait_ms(10);
+    pmw3389_set_cpi(PMW3389_CPI);
+
+    wait_ms(1);
+
+    spi_write_adv(REG_Config2, 0x00);
+
+    spi_write_adv(REG_Angle_Tune, constrain(ROTATIONAL_TRANSFORM_ANGLE, -30, 30));
```
Comment (drashna) [inline drivers/sensors/pmw3389.c:194]: This should be changed to -127 to 127. 
```suggestion
    spi_write_adv(REG_Angle_Tune, constrain(ROTATIONAL_TRANSFORM_ANGLE, -127, 127));
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: drivers/sensors/pmw3389.c (hunk 6 above)
Comment (Alabastard-64) [reply drivers/sensors/pmw3389.c:194]: hhhhmmm I didn't see anything in the docs suggesting that changed and from the 3360 docs it is clearly -30 to 30
image

Comment (drashna) [reply drivers/sensors/pmw3389.c:194]: The same info is in the 3360 datasheet.  But I can verify that it allows the full int8_t range for this register on the 3360, and I'm guessing that it's the same here.

Comment (Alabastard-64) [reply drivers/sensors/pmw3389.c:194]: interesting I wonder what happens to values above 30 and below -30 as based on the hex values shown in that sheet they are unsupported (maybe just defaults to 30 and -30)
by the way based on those hex values it is expecting the two's complement of the negative values do we need any special handling for that?

Comment (drashna) [reply drivers/sensors/pmw3389.c:194]: On the 3360, it "just works", and respects the higher values. 

And no, nothing needs to be done to convert it.

Comment (Alabastard-64) [reply drivers/sensors/pmw3389.c]: huh rereading the documentation nowhere is it stated that 30 and -30 are the maximums...
Alright change is committed as requested

File: quantum/pointing_device_drivers.c [hunk 7]
```diff
@@ -244,6 +244,44 @@ const pointing_device_driver_t pointing_device_driver = {
...
+        mouse_report.x = constrain_hid(data.dx);
+        mouse_report.y = constrain_hid(data.dy);
+    }
+
+    return mouse_report;
+}
+
+// clang-format off
+const pointing_device_driver_t pointing_device_driver = {
+    .init       = pmw3389_init,
```
Comment (drashna) [inline quantum/pointing_device_drivers.c:279]: Sorry, this change should be reverted, or the pmw3389_init function changed to void, rather than bool. (ideally, reverted).

```suggestion
    .init       = init,
```

File: quantum/pointing_device_drivers.c [hunk 8]
```diff
@@ -244,6 +244,44 @@ const pointing_device_driver_t pointing_device_driver = {
     .set_cpi    = pmw3360_set_cpi,
     .get_cpi    = pmw3360_get_cpi
 };
+
+// clang-format on
+#elif defined(POINTING_DEVICE_DRIVER_pmw3389)
+
+
```
Comment (drashna) [inline quantum/pointing_device_drivers.c:251]:
```suggestion
static void init(void) { pmw3389_init(); }

```

File: common_features.mk (hunk 2 above)
Comment (Alabastard-64) [reply common_features.mk:607]: huh wasn't able to commit this in the default way so I have added that line back in
(this was some change on the development branch that I accidentally carried foreword my bad)

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by Alabastard-64** [comment]
added in a small fix for some inconsistent behavior with the cpi settings when swapping between settings seems writing to the top byte first fixes it

**Comment by Alabastard-64** [comment]
Also added a small fix to the write timing checking the docs it seems that the timing for the pmw3360 is still set to the timing of the adns9800 and I carried that foreword to the PMW3389.
I suggest fixing the timing for the PMW3360 (out of scope for this PR though)

*Note the current timing seems mostly functional but it may have been why I was getting weird behavior when writing the CPI settings sometimes (this may have only shown up as an issue when trying to write to two different registers back to back)

**Review by drashna — APPROVED** [review]

**Comment by drashna** [comment]
Yeah, a separate PR for the pmw3360 would be best. Keep things separate.

File: common_features.mk [hunk 9]
```diff
@@ -604,7 +607,7 @@ ifeq ($(strip $(UNICODE_ENABLE)), yes)
...
+	OPT_DEFS += -DUNICODE_COMMON_ENABLE
```
Comment (fauxpark) [inline common_features.mk:610]:
```suggestion
    OPT_DEFS += -DUNICODE_COMMON_ENABLE
```

File: common_features.mk [hunk 10]
```diff
@@ -144,6 +144,9 @@ ifeq ($(strip $(POINTING_DEVICE_ENABLE)), yes)
         else ifeq ($(strip $(POINTING_DEVICE_DRIVER)), pmw3360)
             OPT_DEFS += -DSTM32_SPI -DHAL_USE_SPI=TRUE
             QUANTUM_LIB_SRC += spi_master.c
+	else ifeq ($(strip $(POINTING_DEVICE_DRIVER)), pmw3389)
```
Comment (fauxpark) [inline common_features.mk:147]:
```suggestion
        else ifeq ($(strip $(POINTING_DEVICE_DRIVER)), pmw3389)
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by Alabastard-64** [comment]
All changes applied as requested let me know if anything else is needed

File: drivers/sensors/pmw3389.c [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+#define REG_Revision_ID                0x01
+#define REG_Motion                     0x02
+#define REG_Delta_X_L                  0x03
+#define REG_Delta_X_H                  0x04
+#define REG_Delta_Y_L                  0x05
+#define REG_Delta_Y_H                  0x06
+#define REG_SQUAL                      0x07
+#define REG_RawData_Sum                0x08
+#define REG_Maximum_RawData            0x09
+#define REG_Minimum_RawData            0x0A
```
Comment (uqs) [inline drivers/sensors/pmw3389.c:38]: could you make the hex codes lowercase? That would allow for nicer vimdiff (or the like) between adns and pmw drivers

File: drivers/sensors/pmw3389.c [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+#define REG_LiftCutoff_Cal2            0x65
+#define REG_LiftCutoff_Cal_Timeout     0x71
+#define REG_LiftCutoff_Cal_Min_Length  0x72
+#define REG_PWM_Period_Cnt             0x73
+#define REG_PWM_Width_Cnt              0x74
+
+// clang-format on
+
+#ifndef MAX_CPI
+#    define MAX_CPI 0x013F  // limits to 0--319, should be max cpi/50 -1
```
Comment (uqs) [inline drivers/sensors/pmw3389.c:85]: I found this comment very cryptic and have changed it for pmw3360 in https://github.com/uqs/qmk_firmware/commit/ddca5df530b17a36ac22ea0ea447e8244bd2d025

Did you find then min_cpi that the sensor supports? Can we assume its 50?

File: drivers/sensors/pmw3389.c [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+#endif
+
+bool _inBurst = false;
+
+#ifdef CONSOLE_ENABLE
+void print_byte(uint8_t byte) { dprintf("%c%c%c%c%c%c%c%c|", (byte & 0x80 ? '1' : '0'), (byte & 0x40 ? '1' : '0'), (byte & 0x20 ? '1' : '0'), (byte & 0x10 ? '1' : '0'), (byte & 0x08 ? '1' : '0'), (byte & 0x04 ? '1' : '0'), (byte & 0x02 ? '1' : '0'), (byte & 0x01 ? '1' : '0')); }
+#endif
+#define constrain(amt, low, high) ((amt) < (low) ? (low) : ((amt) > (high) ? (high) : (amt)))
+
+bool spi_start_adv(void) {
```
Comment (uqs) [inline drivers/sensors/pmw3389.c:95]: would be great to keep the names similar between adns and the pmw3360. You might not have seen PR#15559 yet?

File: drivers/sensors/pmw3389.c [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+    wait_us(145);
+    spi_stop();
+    return status;
+}
+
+uint8_t spi_read_adv(uint8_t reg_addr) {
+    spi_start_adv();
+    // send adress of the register, with MSBit = 0 to indicate it's a read
+    spi_write(reg_addr & 0x7f);
+
```
Comment (uqs) [inline drivers/sensors/pmw3389.c:129]: I'm half certain that this misses a tRAD sleep here.

Btw, where did you find the datasheet? I only found a 20 page one, not the 57 page one like for the 3360

**Review by uqs — COMMENTED** [review]
Thanks for these changes, I was really wondering how the CPI settings could work, good to know that indeed the pmw3360 settings for that don't work.

That said, they are really sooooo similar, that adding a copy-pasta of the existing driver doesn't sound too good. We could add 1 or 2 ifdefs to the existing one and it would work just fine.

My worry is that bug fixes (like the missing tSRAD sleep that I'm adding) would get done only for 1 driver and the other one would lag ...

File: drivers/sensors/pmw3389.c (hunk 11 above)
Comment (Alabastard-64) [reply drivers/sensors/pmw3389.c]: ah yeah uppercase is my preference but I should have followed the convention of other drivers

File: drivers/sensors/pmw3389.c (hunk 12 above)
Comment (Alabastard-64) [reply drivers/sensors/pmw3389.c]: as you mentioned nothing in the data sheet explicitly states the min cpi setting I guess we can infer that it is 50 I will do some tests to confirm with this along with testing the tSRAD addition (see below)
I agree with improving the comment I added one that follows your format

File: drivers/sensors/pmw3389.c (hunk 14 above)
Comment (Alabastard-64) [reply drivers/sensors/pmw3389.c:130]: not sure that you need to explicitly wait for tSRAD is this not handled by the spi read? as the read will not finish until the byte is received correct? (which happens after the tSRAD delay)
I am assuming the 20 page document that you are referring to is the same as the one I am using, I just inferred from the changes to the registers and the pmw3360 documentation

File: drivers/sensors/pmw3389.c (hunk 13 above)
Comment (Alabastard-64) [reply drivers/sensors/pmw3389.c:96]: names updated in pmw3389.h and pmw3389.c
missed that PR should be up to date now.

**Comment by Alabastard-64** [comment]
All requested changes have been made. Currently untested will update once I have tested everything works.

I agree about the copy-pasta it does add to the work of maintaining these drivers and merging them would be ideal.
Register definitions get a bit messy as I think we should try and match pixarts docs and they renamed a couple of them and moved some other registers around. Also I think maybe the driver name and shared functions should be given a more generic name such as pmw33xx or pmw.

Please let me know if you think this would work better and I would be happy to submit another PR with the driver added in this way

**Comment by Alabastard-64** [comment]
Upon review the changes in dev branch- the files I am touching in this PR have changed considerably in the development branch since I opened the PR I will need to rebase, update my changes, and test properly before this could be ready for further review and merging.
will need to get back to this in a week once I have time so I would suggest not merging until after I update this properly.

**Comment by Alabastard-64** [comment]
Whew well it seems I borked this PR, sorry about that I have some inexperience with Git/github still.

I have finished adding everyone's suggestions, updated my local repo to the latest dev branch, and tested with a keyboard with the PMW3389.  Everything is working as expected (even better than before in terms of reads).
I will create a new PR with the proper update from a new fork.
